src/solver/algorithm/strong/puzzle.rs

```diff
@@ -9,95 +9,233 @@ use anyhow::{Context, Result};
+fn update_db_record<D>(
+    db: &mut D,
+    state: State,
+    utility: SimpleUtility,
+    remoteness: Remoteness,
+    child_count: ChildCount,
+) -> Result<()>
 where
-    G: DTransition<State> + Bounded<State> + SimpleSum<N>,
-    D: KVStore<RecordBuffer>,
+    D: KVStore,
 {
+    let mut buf = RecordBuffer::from(db.get(state).unwrap())
+        .context("Failed to create record for middle state")?;
+    buf.set_utility([utility])
+        .context("Failed to set utility for state.")?;
+    buf.set_remoteness(remoteness)
+        .context("Failed to set remoteness for state.")?;
+    buf.set_child_count(child_count)
+        .context("Failed to set child count for state.")?;
+    db.put(state, &buf);
 
+    Ok(())
 }
 
-fn discover_end_states<G, D>(db: &mut D, game: &G) -> Vec<State> 
+fn discover_child_counts<G, D>(db: &mut D, game: &G) -> Result<Vec<State>>
 where
-    G: DTransition<State> + Bounded<State> + SimpleSum<N>,
-    D: KVStore<RecordBuffer>,
+    G: DTransition<State>
+        + Bounded<State>
+        + ClassicPuzzle
+        + Extensive<1>
+        + Game,
+    D: KVStore,
 {
-    let visited = HashSet::new();
-    let end_states = Vec::new();
-
-    discover_end_states(db, game, game.start(), visited, end_states);
+    let mut end_states = Vec::new();
+    discover_child_counts_helper(db, game, game.start(), &mut end_states)?;
 
-    end_states
+    Ok(end_states)
 }
 
-fn discover_end_states_helper<G, D>(db: &mut D, game: &G, state: State, visited: HashSet<State>, end_states: Vec<State>)
+/// Adds child counts for each position to the database
+/// Also returns a vector of all primitive positions
+fn discover_child_counts_helper<G, D>(
+    db: &mut D,
+    game: &G,
+    state: State,
+    end_states: &mut Vec<State>,
+) -> Result<()>
 where
-    G: DTransition<State> + Bounded<State> + SimpleSum<N>,
-    D: KVStore<RecordBuffer>,
+    G: DTransition<State>
+        + Bounded<State>
+        + ClassicPuzzle
+        + Extensive<1>
+        + Game,
+    D: KVStore,
 {
-    visited.insert(state);
+    let child_count = game.prograde(state).len() as ChildCount;
 
-    if game.end(state) {
-        end_states.insert(state);
+    if child_count == 0 {
+        end_states.push(state);
     }
 
-    for child in game.prograde(state) {
-        if !visted.contains(child) {
-            discover_end_states(db, game, child, visited, end_states);
+    // Initialize all utilies to draw; any utilities not set by the end must be
+    // a drawn position
+    let mut buf =
+        RecordBuffer::new(1).context("Failed to create record for state")?;
+    buf.set_utility([SimpleUtility::DRAW])
+        .context("Failed to set remoteness for state")?;
+    buf.set_child_count(child_count)
+        .context("Failed to set child count for state.")?;
+    db.put(state, &buf);
+
+    // We need to check both prograde and retrograde; consider a game with 3
+    // nodes where 0-->2 and 1-->2. Then, starting from node 0 with only
```

Comment:
Try to use mkdown formatting for expressing things like `0 --> 2`. Again, if you don't do this it gets rendered weirdly in the docs.